core/src/limited/vec.rs

```diff
@@ -0,0 +1,338 @@
+    ///
+    /// Returns `Ok(())` if the vector can store such number
+    /// of elements and `Err(LimitedVecError)` otherwise.
+    const fn validate_len(len: usize) -> Result<(), LimitedVecError> {
+        if len <= N {
+            Ok(())
+        } else {
+            Err(LimitedVecError)
+        }
+    }
+
+    /// Constructs a new, empty `LimitedVec<T>`.
+    pub const fn new() -> Self {
+        Self(vec![])
+    }
+
+    /// Creates a new limited vector with elements
+    /// initialized with [`Default::default`].
+    pub fn repeat(value: T) -> Self
+    where
+        T: Clone,
+    {
+        Self(vec![value; N])
+    }
+
+    /// Creates a new limited vector with given
+    /// length by repeatedly cloning a value.
+    pub fn try_repeat(value: T, len: usize) -> Result<Self, LimitedVecError>
+    where
+        T: Clone,
+    {
+        Self::validate_len(len).map(|_| Self(vec![value; len]))
+    }
+
+    /// Extends the vector to its limit by
+    /// repeatedly adding a value.
+    pub fn extend_with(&mut self, value: T)
+    where
+        T: Clone,
+    {
+        self.0.resize(N, value)
+    }
+
+    /// Appends a value to the end of the vector.
+    pub fn try_push(&mut self, value: T) -> Result<(), LimitedVecError> {
+        Self::validate_len(self.len() + 1)?;
+
+        self.0.push(value);
+        Ok(())
+    }
+
+    /// Appends values from slice to the end of vector.
+    pub fn try_extend_from_slice(&mut self, values: &[T]) -> Result<(), LimitedVecError>
+    where
+        T: Clone,
+    {
+        let new_len = self
+            .len()
+            .checked_add(values.len())
+            .ok_or(LimitedVecError)?;
+        Self::validate_len(new_len)?;
+
+        self.0.extend_from_slice(values);
+        Ok(())
+    }
+
+    /// Returns a slice reference to the vector contents.
+    pub fn as_slice(&self) -> &[T] {
+        self
+    }
+
+    /// Clones the limited vector into `Vec<T>`.
+    pub fn to_vec(&self) -> Vec<T>
+    where
+        T: Clone,
+    {
+        self.0.clone()
+    }
+
+    /// Converts the limited vector into its inner `Vec<T>`.
+    pub fn into_vec(self) -> Vec<T> {
+        self.0
+    }
+}
+
+impl<T, const N: usize> fmt::Display for LimitedVec<T, N>
+where
+    [T]: AsRef<[u8]>,
+{
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        let bytes = ByteSliceFormatter::Dynamic(self.0.as_slice().as_ref());
+
+        let precision = f.precision();
+        let sign_plus = f.sign_plus();
+
+        let fmt_bytes: fn(&mut Formatter, fmt::Arguments) -> _ = if f.alternate() {
+            |f, bytes| write!(f, "LimitedVec({bytes})")
+        } else {
+            |f, bytes| write!(f, "{bytes}")
+        };
+
+        if let Some(precision) = precision {
+            fmt_bytes(f, format_args!("{bytes:.precision$}"))
+        } else if sign_plus {
+            fmt_bytes(f, format_args!("{bytes}"))
+        } else {
+            fmt_bytes(f, format_args!("{bytes:.8}"))
+        }
```

Comment:
Can be simplified like this
```suggestion
        let bytes = ByteSliceFormatter::Dynamic(self.0.as_slice().as_ref());

        let precision = f.precision();
        let sign_plus = f.sign_plus();

        let bytes = if let Some(precision) = precision {
            format_args!("{bytes:.precision$}")
        } else if sign_plus {
            format_args!("{bytes}")
        } else {
            format_args!("{bytes:.8}")
        };
        
        if f.alternate() {
            write!(f, "LimitedVec({bytes})")
        } else {
            write!(f, "{bytes}")
        }
```

Reply:
No, it can not.

https://doc.rust-lang.org/std/macro.format_args.html#lifetime-limitation

Reply:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=5c4e74ebb93aa581e1b59e32b2899318